code/go/0chain.net/smartcontract/storagesc/readpool.go

```diff
@@ -84,380 +98,156 @@ func toJson(val interface{}) string {
 	return string(b)
 }
 
-func (rp *readPool) moveToBlobber(sscKey, allocID, blobID string,
-	sp *stakePool, now common.Timestamp, value state.Balance,
+func (rp *readPool) moveToBlobber(allocID, blobID string,
+	sp *stakePool, value state.Balance, isOwner bool,
 	balances cstate.StateContextI) (resp string, err error) {
 
-	var cut = rp.blobberCut(allocID, blobID, now)
-
-	if len(cut) == 0 {
-		return "", fmt.Errorf("no tokens in read pool for allocation: %s,"+
-			" blobber: %s", allocID, blobID)
-	}
-
 	// all redeems to response at the end
 	var redeems []readPoolRedeem
-	var moved state.Balance = 0
-	var torm []*allocationPool // to remove later (empty allocation pools)
-	for _, ap := range cut {
-		if value == moved {
-			break // all required tokens has moved to the blobber
-		}
-		var bi, ok = ap.Blobbers.getIndex(blobID)
-		if !ok {
-			continue // impossible case, but leave the check here
-		}
-		var (
-			bp   = ap.Blobbers[bi]
-			move state.Balance
-		)
-		if value >= bp.Balance {
-			move, bp.Balance = bp.Balance, 0
-		} else {
-			move, bp.Balance = value, bp.Balance-value
-		}
-
-		ap.Balance -= state.Balance(value)
-
-		redeems = append(redeems, readPoolRedeem{
-			PoolID:  ap.ID,
-			Balance: move,
-		})
-
-		moved += move
-		if bp.Balance == 0 {
-			ap.Blobbers.removeByIndex(bi)
-		}
-		if ap.Balance == 0 {
-			torm = append(torm, ap) // remove the allocation pool later
-		}
+	var moved state.Balance
+	currentBalance := rp.OwnerBalance
+	ownership := "owner"
+	if !isOwner {
+		currentBalance = rp.VisitorBalance
+		ownership = "visitor"
+	}
+	if value >= currentBalance {
+		return "", fmt.Errorf("not enough tokens in %s read pool for "+
+			"allocation: %s, blobber: %s", ownership, allocID, blobID)
+	} else {
+		moved, currentBalance = value, currentBalance-value
 	}
 
-	if moved < value {
-		return "", fmt.Errorf("not enough tokens in read pool for "+
-			"allocation: %s, blobber: %s", allocID, blobID)
+	redeems = append(redeems, readPoolRedeem{
```

Comment:
can `redeems` contain more than one item?

Reply:
no, but the response expects []readPoolRedeem, so we have to return a single value to confirm to legacy codes, once we find all the places where this value is called, we can change the array to a single item (structure).